update post.md

kdpanw opened the pull request:
See Jira for more info : https://redlock.atlassian.net/browse/PCSUP-6947 . The collections POST endpoint does not instantiate correctly VIA api call . You must instantiate all values applicable as a * wildcard, in order for them to be added. They do not add by default.

kdpanw committed:
update post.md

See Jira for more info : https://redlock.atlassian.net/browse/PCSUP-6947 . The collections POST endpoint does not instantiate correctly VIA api call . You must instantiate all values applicable as a * wildcard, in order for them to be added. They do not add by default.

iansk commented:
@ssugandh Please validate the current content, and fix if necessary.

ssugandh commented on api/descriptions/collections/post.md:
````diff
@@ -23,5 +23,6 @@ $ curl 'https://<CONSOLE>/api/v<VERSION>/collections' \
 }'
 ```
 Any resource left unspecified, such as `hosts`, `functions`, or `clusters`, is assigned a wildcard by default.
+Resources that are not explicitly specified need to have a `*` wildcard applied to them in order to instantiate.
````
The statement at line number 26 suggests some user action for adding a wildcard '*'. But, it's automatically gets added for instantiation as mentioned in line number 25. We can rephrase as "To instantiate a resource object, a wildcard '*' is assigned by default if the resource is left unspecified."

kdpanw replied:
Hi, 
   If you create a collection via the API and do not specify a wildcard, it will not be added. I can demo this is necessary. That is why I suggest we remove the line saying it is added by default, as it is not.

ssugandh replied:
Okay, I maybe wrong but during my testing when I created a collection via API, the wildcards got added as mentioned.

kdpanw replied:
I am not sure what operation you performed or what test you ran to do this, it's a specific use-case. This may be easier to sync on with a quick huddle over slack if you'd like. Here, @ssugandh , take a quick look at my repo on this JIRA: https://redlock.atlassian.net/browse/PCSUP-6947 . I have some images in there that should shed some light on this.

ssugandh replied:
I see that. The issue seems to be true. I verified as suggested in the Jira.

When I create a collection via API endpoint, the wildcards are added or atleast it seems to be added on the UI. I verified with the UI. But, when I use that same collection to create a rule, it throws that error like it's mentioned in the Jira. Unless, I pass a wildcard explicitly through either POST or PUT, it's treated as empty.

iansk replied:
Excellent work, @ssugandh 

@kdpanw Does this cover the full extent of the issue?

iansk replied:
This line has been updated. Resolving.

ssugandh commented on api/descriptions/collections/post.md:
```diff
@@ -23,5 +23,6 @@ $ curl 'https://<CONSOLE>/api/v<VERSION>/collections' \
+Resources that are not explicitly specified need to have a `*` wildcard applied to them in order to instantiate.
+
 
-**Note:** No response will be returned upon successful execution.
```
This is still valid. Not sure, why this got removed.

kdpanw replied:
There are situations where we receive no response due to perhaps a 404/400 error but the cURL has no response shown, and this indicates that it is successful. However, it is a failure. Running cURL -vvv on your API calls can show this in some aspects.

kdpanw replied:
If you are certain that this functionality is behaving as expected, perhaps it is some bug with cURL or some other variable, and I can research that and see what I find and create a different PR/Report for that aspect.

ssugandh replied:
True, but we have several other instances similar to this. Shall we remove such statements from other API descriptions too?

kdpanw replied:
I personally believe we should. I do not think there is any added value from having this statement, as it can confuse customers. It also is a blanket statement that doesn't necessarily cover every single use-case. Also, product team (I assume) is not keeping up with that as some sort of standard. What I do know, is that I have seen this occur where we have no response, but it's not completing the API call. It's getting a 404. I think that we get more benefit from removing this statement, than we do from having it in the docs. Generally, customers using the API are well-off enough to navigate around themselves and will know whether or not it works.

iansk replied:
@ssugandh What is the response code when a collection is successfully created? 200?

ssugandh replied:
@iansk : There's no response for a successful execution. I have retained the previous line with a caveat to verify with the Console UI.

ssugandh committed:
Update post.md

Update the description and example.

iansk commented on api/descriptions/collections/post.md:
```diff
@@ -1,4 +1,4 @@
-Creates a new collection.
+Creates a new collection. You need to either specify a value or initialize the resouces in the endpoint with a "*" wildcard to be able to use the collection.
```
@ssugandh Couple of things:

1. Please use semantic lines breaks (i.e. one sentence per line). See the guidance here: https://sembr.org/

2. Please update this sentence to: 

   ```
   You must specify a value for all resources in the collection, even if it's just a wildcard `*`.
   ```

3. Add a new paragraph that describes why it's important to specify a value for all resources, and what happens if you don't do that.

   ```
   If you create a collection and specify values for only some resources, then ... <WHAT'S THE OUTCOME, WHY THIS IS IMPORTANT, HOW YOU CAN TELL YOU DID SOMETHING WRONG, HOW TO VALIDATE, ETC>
   ```

iansk commented on api/descriptions/collections/post.md:
```diff
@@ -8,20 +8,30 @@ To invoke this endpoint in the Console UI:
 ### cURL Request
 
-Refer to the following example cURL command that creates a new collection named `my-collection` and captures all container images named `ubuntu:18.04`:
+Refer to the following example cURL command that creates a new collection named `my-collection`, specifies a HEX color value of #AD3C21, and captures all container images named `ubuntu:18.04`:
```
@ssugandh What happens if you don't specify a color? Does the system randomly assign the collection a unique color? Please test it in your lab setup. If it does, then remove this param from the example curl command.

I want our example curl commands to show only the **minimum params** do required to do something **useful in the real world**. (Sometimes our API lets you specify even fewer params then we show in our curl commands, but the created object doesn't do anything)

iansk replied:
@ssugandh The reason I want our curl commands specified this way is because our OpenAPI spec doesn't declare which params are optional and which are required. As such, I want our curl commands to act as a stand-in for this missing info.

Customers can then take our "minimally useful" curl command, and build on it to implement their own use case.

ssugandh replied:
@iansk : I've verified that only three resources are manadatory to use the collection, name, images, and labels. I have described about it in the description. I've retained the example with some resources just as a reference. Let me know if you still think otherwise, I can remove the non-mandatory resources from the example. 

In my other experiences to validate cURL example for creating users, I struggled to get the exact names of the property "role". I had to create users in the UI and use GET to understand all the exact values of the "role" used in the JSON file.

For color, should we not update about the random color in the parameter description?

iansk commented on api/descriptions/collections/post.md:
````diff
@@ -8,20 +8,30 @@ To invoke this endpoint in the Console UI:
 }'
 ```
-Any resource left unspecified, such as `hosts`, `functions`, or `clusters`, is assigned a wildcard by default.
+**Note:** Resources that are not explicitly specified need to have a `*` wildcard applied to them in order to instantiate as mentioned in the example.
````
@ssugandh Let's remove this line. You're already adding a lot of info about this issue at the beginning of the endpoint description. No need to repeat it again here.

ssugandh committed:
Update post.md

Updated description. I tested with all resources and only three are mandatory for initialisation to use the collection.

ssugandh commented:
@iansk : Updated all.

kdpanw commented:
@iansk @Pubs-MV Hey team, did this ever get merged?

eric-bischoff reviewed (approved):
This is good to be merged.

Pubs-MV commented:
@ssugandh can you please confirm  that this does not conflict and can be merged?

ssugandh commented:
Yes, this can be merged. I don't have the merge permissions.

ssugandh commented:
@iansk and @Pubs-MV : Can you please merge this? I don't have merge permissions.
